### PR TITLE
[SPARK-24991][SQL] use InternalRow in DataSourceWriter

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

A follow up of #21118

Since we use `InternalRow` in the read API of data source v2, we should do the same thing for the write API.

## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by holdensmagicalunicorn** [comment]
@cloud-fan, thanks! I am a bot who has found some folks who might be able to help with the review:@gatorsmile, @zsxwing and @tdas

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala [hunk 1]
```diff
@@ -89,8 +89,7 @@ class RateStreamContinuousReader(options: DataSourceOptions) extends ContinuousR
         start.runTimeMs,
         i,
         numPartitions,
-        perPartitionRate)
-        .asInstanceOf[InputPartition[InternalRow]]
+        perPartitionRate): InputPartition[InternalRow]
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: This is needed to make it compile, but at least we don't need to do cast.

**Comment by cloud-fan** [comment]
cc @rdblue @jose-torres @gatorsmile

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: Sounds good to me.

**Comment by rdblue** [comment]
+1

**Comment by jose-torres** [comment]
lgtm

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java [hunk 2]
```diff
@@ -50,4 +50,15 @@
    *                this ID will always be 0.
    */
   DataWriter<T> createDataWriter(int partitionId, long taskId, long epochId);
+
+  /**
+   * When true, Spark will reuse the same data object instance when sending data to the data writer,
+   * for better performance. Data writers should carefully handle the data objects if it's reused,
+   * e.g. do not buffer the data objects in a list. By default it returns false for safety, data
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: @rdblue did you hit this problem in iceberg?

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala [hunk 3]
```diff
@@ -113,11 +109,15 @@ object DataWritingSparkTask extends Logging {
...
     val epochId = Option(context.getLocalProperty(MicroBatchExecution.BATCH_ID_KEY)).getOrElse("0")
     val dataWriter = writeTask.createDataWriter(partId, taskId, epochId.toLong)
+    val copyIfNeeded: InternalRow => InternalRow =
+      if (writeTask.reuseDataObject()) identity else _.copy()
 
     // write the data and commit this writer.
     Utils.tryWithSafeFinallyAndFailureCallbacks(block = {
       while (iter.hasNext) {
-        dataWriter.write(iter.next())
+        // Internally Spark reuse the same UnsafeRow instance when producing output rows, here we
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala]: nit: reuse => reuses

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java [hunk 4]
```diff
@@ -50,4 +50,15 @@
    *                this ID will always be 0.
    */
   DataWriter<T> createDataWriter(int partitionId, long taskId, long epochId);
+
+  /**
+   * When true, Spark will reuse the same data object instance when sending data to the data writer,
+   * for better performance. Data writers should carefully handle the data objects if it's reused,
```
Comment (gengliangwang) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: Nit: `if it's reused` the `it` here is ambiguous. Maybe change to `if they are reused`?

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java (hunk 2 above)
Comment (gengliangwang) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: nit: `By default it returns false, data sources` => `By default the method returns false. Data sources`

**Review by gengliangwang — APPROVED** [review]
LGTM otherwise.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java (hunk 2 above)
Comment (rdblue) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: No, Iceberg assumes that data objects are reused.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java [hunk 5]
```diff
@@ -50,4 +50,15 @@
...
+   * e.g. do not buffer the data objects in a list. By default it returns false for safety, data
+   * sources can override it if their data writers immediately write the data object to somewhere
+   * else like a memory buffer or disk.
+   */
+  default boolean reuseDataObject() {
```
Comment (rdblue) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: I don't think this should be added in this commit. This is to move to `InternalRow` and should not alter the API. I'm fine documenting this, but writers are responsible for defensive copies if necessary. This default is going to cause sources to be slower and I don't think it is necessary for implementations that aren't tests buffering data in memory.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 1 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: This should be in a separate commit. I didn't notice yesterday that this is for the writer until it was linked from the other issue. I think this change needs to get in, but it should not be mixed into changes for the write path.

**Comment by rdblue** [comment]
I'm changing my +1 to -1 because read-side changes are mixed in and because copies are the responsibility of data sources if they buffer and hold references to earlier rows.

**Comment by cloud-fan** [comment]
@rdblue I have documented the object reuse behavior and ask data source to handle it, please take a look, thanks!

**Comment by cloud-fan** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala [hunk 6]
```diff
@@ -89,7 +89,8 @@ class RateStreamContinuousReader(options: DataSourceOptions) extends ContinuousR
         start.runTimeMs,
         i,
         numPartitions,
-        perPartitionRate): InputPartition[InternalRow]
+        perPartitionRate)
+        .asInstanceOf[InputPartition[InternalRow]]
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: Why is this cast necessary?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala [hunk 7]
```diff
@@ -44,16 +46,16 @@ class MemorySinkV2Suite extends StreamTest with BeforeAndAfter {
     val writer = new MemoryStreamWriter(sink, OutputMode.Append(), new StructType().add("i", "int"))
     writer.commit(0,
       Array(
-        MemoryWriterCommitMessage(0, Seq(InternalRow(1), InternalRow(2))),
-        MemoryWriterCommitMessage(1, Seq(InternalRow(3), InternalRow(4))),
-        MemoryWriterCommitMessage(2, Seq(InternalRow(6), InternalRow(7)))
+        MemoryWriterCommitMessage(0, Seq(Row(1), Row(2))),
```
Comment (rdblue) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala]: Why was this changed back to Row?

**Comment by rdblue** [comment]
@cloud-fan, thanks for documenting the behavior and removing the default copy. I had a couple of questions, but I think it is close.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: This is to address your comments: do not mix read-side changes. So I reverted it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala]: Now the `DataWriter` needs to copy the input row before buffering it, which can be done by the `RowEncoder` when converting `InternalRow` to `Row`. Then the write message carries `Row`s to the driver side.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala (hunk 6 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala]: Okay, sorry about that. I should have looked at the whole diff.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala]: Why not use InternalRow.copy? I'd rather keep the update to InternalRow, but as long as the tests pass I wouldn't block this.

**Comment by rdblue** [comment]
+1 when tests are passing.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java [hunk 8]
```diff
@@ -33,7 +33,10 @@
 public interface DataWriterFactory<T> extends Serializable {
 
   /**
-   * Returns a data writer to do the actual writing work.
+   * Returns a data writer to do the actual writing work. Note that, Spark will reuse the same data
+   * object instance when sending data to the data writer, for better performance. Data writers
+   * are responsible for defensive copies if necessary, e.g. copy the data before buffer it in a
+   * list.
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: nit: the description about defensive copied in data writers, may be put in `DataWriter`.

**Comment by viirya** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala]: because the memory sink needs `Row`s at the end. Instead of collecting `InternalRow`s via copy and then convert to `Row`s, I think it's more efficient to collect `Row`s directly.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java (hunk 8 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataWriterFactory.java]: I'll fix it in my next PR, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala (hunk 7 above)
Comment (rdblue) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/MemorySinkV2Suite.scala]: Sounds good. Thanks for explaining your rationale.
